streams/src/test/java/org/apache/kafka/streams/integration/ResetPartitionTimeIntegrationTest.java

```diff
@@ -0,0 +1,279 @@
+        );
+    }
+
+    public ResetPartitionTimeIntegrationTest(final boolean eosEnabled) {
+        this.eosEnabled = eosEnabled;
+    }
+
+    @Test
+    public void testPartitionTimeAfterKStreamReset() throws InterruptedException {
+        final String testId = "-shouldRecoverPartitionTimeAfterReset";
+        final String appId = getClass().getSimpleName().toLowerCase(Locale.getDefault()) + testId;
+        final String input = "input" + testId;
+        final String storeName = "counts";
+        final String outputRaw = "output-raw" + testId;
+
+        // create multiple partitions as a trap, in case the buffer doesn't properly set the
+        // partition on the records, but instead relies on the default key partitioner
+        cleanStateBeforeTest(CLUSTER, 2, input, outputRaw);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+        final KStream<String, Long> valueCounts = builder
+            .stream(
+                input,
+                Consumed.with(STRING_SERDE, STRING_SERDE))
+            .groupByKey()
+            .count(Materialized.<String, Long, KeyValueStore<Bytes, byte[]>>as(storeName).withCachingDisabled())
+            .toStream();
+
+        // expect all post-suppress records to keep the right input topic
+        final MetadataValidator metadataValidator = new MetadataValidator(input);
+
+        valueCounts
+            .transform(metadataValidator)
+            .to(outputRaw, Produced.with(STRING_SERDE, Serdes.Long()));
+
+        final Properties streamsConfig = new Properties();
+        streamsConfig.put(StreamsConfig.DEFAULT_TIMESTAMP_EXTRACTOR_CLASS_CONFIG, MaxTimestampExtractor.class);
+        streamsConfig.put(StreamsConfig.APPLICATION_ID_CONFIG, appId);
+        streamsConfig.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfig.put(StreamsConfig.POLL_MS_CONFIG, Integer.toString(DEFAULT_TIMEOUT));
+        streamsConfig.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, Integer.toString(DEFAULT_TIMEOUT));
+        streamsConfig.put(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, eosEnabled ? EXACTLY_ONCE : AT_LEAST_ONCE);
+        streamsConfig.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+        
+
+        KafkaStreams driver = getStartedStreams(streamsConfig, builder, true);
+        try {
+            // start sending some records to have partition time committed 
+            produceSynchronouslyToPartitionZero(
+                input,
+                asList(
+                    new KeyValueTimestamp<>("k3", "v3", 5000)
+                )
+            );
+            verifyOutput(
+                outputRaw,
+                asList(
+                    new KeyValueTimestamp<>("k3", 1L, 5000)
+                )
+            );
+            assertThat(lastRecordedTimestamp, is(5000L));
+            lastRecordedTimestamp = -1L;
+            Thread.sleep(1000); // wait for commit to finish
```

Comment:
We don't need to sleep (in general, sleeping is bad practice because it makes test flaky), because when `close()` is called below, it is ensured that offsets are committed.

Reply:
Well, I put the sleep there because otherwise the test (eos enabled case) breaks. 

I have done quite a bit of digging, and it appears what happens is that the committed metadata retrieved is incorrect after the streams restart. I added some debug statements, and the strange thing is though is that committed() doesn't return the right metadata.
I made two calls to metadata() -- this is during initializeTaskTime() -- and the first call returns the incorrect metadata (the result suggests that no OffsetAndMetadata was committed), yet on the second call, it returns the correct metadata.

The sleep() method I put there because it seems that commit() has enough time to actually persist in Kafka log, otherwise, Consumer#committed() returns inconsistent results.

Reply:
That is weird. \cc @guozhangwang How could the happen? If we stop the first instance, the transactions should be committed and afterwards, if we start a new instance, the new consumer client should be able to read the correct offset and metadata. Any idea?